proof-language.md

```diff
@@ -0,0 +1,359 @@
+---
+geometry: margin=2cm
+---
+
+\newcommand{\limplies}{\to}
+\renewcommand{\phi}{\varphi}
+
+
+Matching logic proof format
+===========================
+
+This is a proposed matching logic proof format.
+
+## Relation to Metamath
+
+Note that this language heavily based on the metamath format for compressed proofs.
+Besides being a binary rather than ASCII encoding, there are two major changes.
+
+1.  We use the same mechanism for constructing terms and proofs, allowing better
+    compression.
+2.  We allow emmitting multiple proofs from the same representation. This allows
+    us to represent lemmas and compression uniformly.
+    That is, we can store the entire contents of a single metamath file in this format,
+    rather than just the proof of a single lemma.
+
+## Relation to postfix notation/RPN
+
+The format may be viewed as a generalization
+of postfix notation (incidentally also called Lukaseiwicz notation),
+for an expression computing the conclusion of a proof.
+
+We also allow marking subexpressions (i.e. either proofs or patterns) for reuse. This
+allows us to represent lemmas, and compactly representing configurations and notation.
+
+(This needs input from ZK folk.) Finally, we allow marking constructed proofs
+as public. This serves two purposes. First, it allows us to publish lemmas
+that may be reused in other proofs. Second, it allows us to publish the main
+result proved by the proof.
+
+
+## As a stack machine
+
+It may also be viewed language interpreted using a stack machine.
+This stack machines constructs terms---either patterns or proofs. We treat
+patterns and proofs uniformly---at an abstract level, they are both terms we
+must construct, with particular well-formedness criteria. It is intended for the
+most part, that **the verifier only needs to keep proved conclusions in
+memory**.
+
+
+## Goals:
+
+-   No upfront parsing cost.
+-   Easily verifiable using ZK
+-   Low memory usage.
+-   Composability
+
+
+## Non-goals:
+
+-   Parallization: I have not been thinking explicitly about parallel verification of this format,
+    because I expect parallization to happen by composing units of this format into larger proofs.
+    Even so I think with some additional metadata, we could make
+    this format parallizable.
+
+-   Human readability: We aim to be a simple binary format. To become human
+    readable, we expect the user to run a secondary program to pretty-print the
+    proof. Note that symbols, lemmas, and notation do not have names.
+
+
+Terms
+=====
+
+As mentioned previously, we may view this format as a language for constructing
+terms that meet particular well-formedness criteria. Let us first enumerate
+those terms, in a python-esque pseudocode.
+
+`Term`s are of two types, `Pattern`, and `Proof`.
+Every `Term` has a well-formedness check.
+Every `Proof` has a conclusion.
+
+```python
+abstract class Term:
+    abstract def well_formed:
+        ...
+
+abstract class Pattern(Term):
```

Comment:
Maybe consider adding Contexts that should have the distinguished hole variable check, as we discussed?

Reply:
I'm not sure I recall what we discussed. `MetaVar`s already have an `application_context` field that is used for this check.